core/shared/platform/zephyr/zephyr_socket.c

```diff
@@ -0,0 +1,1049 @@
+        case DNS_EAI_SYSTEM:
+            return errno;
+        default:
+            return EINVAL;
+    }
+}
+
+static int
+is_addrinfo_supported(struct zsock_addrinfo *info)
+{
+    return
+        // Allow only IPv4 and IPv6
+        (info->ai_family == AF_INET || info->ai_family == AF_INET6)
+        // Allow only UDP and TCP
+        && (info->ai_socktype == SOCK_DGRAM || info->ai_socktype == SOCK_STREAM)
+        && (info->ai_protocol == IPPROTO_TCP
+            || info->ai_protocol == IPPROTO_UDP);
+}
+
+static int
+os_socket_setbooloption(bh_socket_t socket, int level, int optname,
+                        bool is_enabled)
+{
+    int option = (int)is_enabled;
+
+    if (zsock_setsockopt(socket->fd, level, optname, &option, sizeof(option))
+        != 0) {
+        return BHT_ERROR;
+    }
+
+    return BHT_OK;
+}
+
+static int
+os_socket_getbooloption(bh_socket_t socket, int level, int optname,
+                        bool *is_enabled)
+{
+    assert(is_enabled);
+
+    int optval;
+    socklen_t optval_size = sizeof(optval);
+
+    if (zsock_setsockopt(socket->fd, level, optname, &optval, optval_size)
+        != 0) {
+        return BHT_ERROR;
+    }
+    *is_enabled = (bool)optval;
+    return BHT_OK;
+}
+
+// Platform API implementation
+int
+os_socket_create(bh_socket_t *sock, bool is_ipv4, bool is_tcp)
+{
+    int af = is_ipv4 ? AF_INET : AF_INET6;
+    // now socket is a struct try  *(sock)->fd
+
+    *(sock) = BH_MALLOC(sizeof(zephyr_handle));
+
+    if (!sock) {
+        return BHT_ERROR;
+    }
+
+    if (is_tcp) {
+        (*sock)->fd = zsock_socket(af, SOCK_STREAM, IPPROTO_TCP);
+    }
+    else {
+        (*sock)->fd =
+            zsock_socket(af, SOCK_DGRAM, IPPROTO_UDP); // IPPROTO_UDP or 0 ?
+    }
+
+    (*sock)->is_sock = true;
+
+    return ((*sock)->fd == -1) ? BHT_ERROR : BHT_OK;
+}
+
+int
+os_socket_bind(bh_socket_t socket, const char *host, int *port)
+{
+    struct sockaddr_storage addr = { 0 };
+    socklen_t socklen;
+    int ret;
+
+    assert(host);
+    assert(port);
+
+    if (!textual_addr_to_sockaddr(host, *port, (struct sockaddr *)&addr,
+                                  &socklen)) {
+        goto fail;
+    }
+
+    // F_SETF_SETFD and FD_CLOEXEC are not defined in zephyr.
+    // SO_LINGER: Socket lingers on close (ignored, for compatibility)
+
+    ret = zsock_bind(socket->fd, (struct sockaddr *)&addr, socklen);
+    if (ret < 0) {
+        goto fail;
+    }
+
+    socklen = sizeof(addr);
+    if (zsock_getsockname(socket->fd, (void *)&addr, &socklen) == -1) {
+        goto fail;
+    }
+
+    if (addr.ss_family == AF_INET) { // addr.sin_family
+        *port = ntohs(((struct sockaddr_in *)&addr)->sin_port);
+    }
+    else {
+#ifdef IPPROTO_IPV6
+        *port = ntohs(((struct sockaddr_in *)&addr)->sin6_port);
+#else
+        goto fail;
+#endif
+    }
+
+    return BHT_OK;
+
+fail:
+    // Close the fd because FD_CLOEXEC isn't implemented.
+    os_socket_close(socket);
```

Comment:
os_socket_close will close the `socket->fd` and BH_FREE(socket), but they are not opened and allocated in this function, seems here we had better just call `os_socket_shutdown` and return BHT_ERROR?

Reply:
Leaving `os_socket_close` here is an error. 

IMO, even placing `os_socket_shutdown` is not necessary because when `bind` fail there is no such thing as shutdown. It should be the user responsability in higher level code to test `bind` return value and handle it accordingly (either `close` or `shutdown`).

eg:
```c
int result = bind(socket, server_addr, addr_len);
if (result == -1)
{
    // handle failure
    close(socket); 
    // or shutdown(socket); 
}
```

Reply:
OK, thanks for the correction.